[WIP] Implement custom Docker container CI via GitHub Actions

carmenbianca opened the pull request:
This PR tries to re-implement the CI functionality that was lost when Travis stopped being used.

It's a heavy WIP that is undertaken as a side project by a newbie cooperator (that's me). Ideally the work done here could be extracted into a template. vertical-cooperative was chosen as a guinea pig because it's small-medium-sized and doesn't have that many external dependencies.

edit (2021-12-15): This uses https://github.com/carmenbianca/cie-odoo-ci as Docker Container now.

carmenbianca commented:
I'm not exactly sure why this isn't working. [oca-ci](https://github.com/OCA/oca-ci) attempts to pip install everything in `setup/`. Because of [setuptools-odoo](https://pypi.org/project/setuptools-odoo/), all the dependencies declared in the modules' `__manifest__.py` files are translated to names of Python modules (e.g., [odoo12-addon-email_template_config](https://pypi.org/project/odoo12-addon-email-template-config/)). These modules are available for installation on PyPI. (Were these packages manually created on PyPI or added by a bot?)

When I do `pip install -e setup/easy_my_coop` locally, everything installs fine, including `odoo12-addon-email_template_config`. When I do `pip install -r test-requirements.txt` using the same file as the one in the CI, I come across a dependency on [odoo12-addon-auth_company_signup](https://github.com/coopiteasy/addons/tree/12.0/auth_company_signup) which is **not** available on PyPI.

Two observations therefore:

1. Although I have no idea why the CI is failing on odoo12-addon-email_template_config, it is assuredly going to fail on odoo12-addon-auth_company_signup, and maybe some other modules.
2. I am doubtful that the method of installing modules via pip/PyPI is sufficiently sturdy. I'm sure it _can be_ sturdy, but we're not exactly using pip installation for Odoo modules anywhere at the moment, to my best understanding.

Which leads me to three possible paths at tackling this problem, keeping in mind a limited understanding of Odoo:

1. Don't use oci-ci. Create something else that stays away from Python packaging. To my understanding this is not desirable because the OCA likes to keep things very consistent.
2. Switch over to Python packaging, wholesale. My presumption here is that the pursuit of adopting Python packaging instead of gitaggregate would solve the things that need solving to make this work. The problem is that this is would be a rather big project. @huguesdk had done some research on this.
3. Keep hacking away at this until it works. Probably.

carmenbianca commented on easy_my_coop_website_portal/readme/DESCRIPTION.rst:
```diff
@@ -3,4 +3,3 @@
 * On page /my/account
    * prevents modifications to "name" and "email"
    * makes "iban", "birthdate_date", "gender" and "lang" compulsory
-* ...
```
@robinkeunen Please verify that this change is correct. pylint was complaining about this line, so I deleted it.

robinkeunen replied:
👌

carmenbianca commented:
Found the culprit for odoo12-addon-email_template_config not installing in the CI, but working just fine locally. In oca-ci's [Dockerfile](https://github.com/OCA/oca-ci/blob/master/Dockerfile):

```dockerfile
# This PEP 503 index uses odoo addons from OCA and redirects the rest to PyPI,
# in effect hiding all non-OCA Odoo addons that are on PyPI.
ENV PIP_INDEX_URL=https://wheelhouse.odoo-community.org/oca-simple-and-pypi
```

carmenbianca commented:
IT WORKS! It just needs a lot of polishing now.

robinkeunen commented:
![](https://media.giphy.com/media/YEL7FJP6ed008/giphy.gif)

@huguesdk @remytms

robinkeunen commented:
Good job @carmenbianca, it's good to have you with us !!

Another thing that might be good to follow is this repo: https://github.com/sbidoul/runboat

carmenbianca committed:
[ADD] Create GitHub Action workflow for pre-commit

This ensures that the entirety of pre-commit tests is run, which catches
errors that are sometimes not caught when only running the tests for a
subset of files.

It would be nicer if there were a separate workflow/job for every
pre-commit test, because that would make the output more easily
parseable by human eyes, but this will probably do.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] Create GitHub Actions workflow for Odoo testing

Most of the functionality rests in
<https://github.com/coopiteasy/cie-odoo-ci>.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented:
@robinkeunen Thanks! :tada:

In re runboat: I'll have to take a long afternoon to catch up on that. I'm not very familiar with Kubernetes, and runboat appears to be doing a lot. Putting it on my backlog somewhere.

Also here some notes to self on things to wrap up before this PR is complete:

- Figure out why the pre-commit workflow is not failing. It appears to be hitting a wrong cache?
- We ultimately probably want to make this compatible for multiple versions of Odoo. Very little needs to change other than:
  - refresh_gitaggregate needs to download the correct repos.yml for that version.
  - The containers need to be tagged correctly, both here and upstream.
- Same as above, but for multiple Python versions.
  - gitaggregate doesn't need altering for this.
  - This workflow would probably need a job matrix.
- Move https://github.com/carmenbianca/cie-odoo-ci to the coopiteasy org (and adjust this workflow).
- We currently depend on the `repos.yml` files in https://gitlab.com/coopiteasy/cie-repositories quite heavily. This probably isn't _ideal_, because whether-or-not these runs succeed depends on upstream configuration to be correct. Not sure how to fix this.
- Do we need this working on GitLab as well? I don't think it would take a lot of work other than (nightmarishly slowly) debugging the CI config, because most of the work is done in the container.
- Verify whether pre-commit correctly covers all needed tests.
  - Is there one for the readme files that I keep forgetting to generate?
- edit: /etc/odoo.cfg is hardcoded and this isn't _great_.
- Probably more.

robinkeunen commented:
The runboat is still in its infancy so no need to follow too closely. I think it'd be a good subject to be aware for the CC of January. It would also be a good entry point if we/you want
- to peak into the oca infrastructure and collaborate there
- to mutualize some infrastructure plumbing with the GRAP / the filament / other coops (we talked about that with [Sylvain Legal](https://github.com/legalsylvain))

carmenbianca commented:
@huguesdk @robinkeunen Ready for review.

Important link where the Dockerfile lives: https://github.com/coopiteasy/cie-odoo-ci

Done:

- pre-commit runs on all files
- Odoo tests run against our test gitaggregate configuration (+ the PR's working directory)
- Odoo tests run against Python 3.6, 3.7, and 3.8.

Not done:

- Odoo tests run against multiple versions of Odoo.
- `addons_path` in `/etc/odoo.cfg` is hardcoded in the Docker container. Not great.
- Figure out on which `repos.yml` files to depend to pull all dependencies.
- I haven't vetted the tests inside of `.pre-commit-config.yaml` for completeness.

`pre-commit` currently fails because of c41d3d8 by @robinkeunen. The Odoo tests fail for reasons I haven't investigated.

robinkeunen commented:
@remytms

carmenbianca commented:
Made obsolete
